File: android/app/src/main/kotlin/run/nuri/getagrip/ui/runner/DebugSummaryPreview.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.
package run.nuri.getagrip.ui.runner

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.safeDrawingPadding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import run.nuri.getagrip.engine.*
import run.nuri.getagrip.runner.MaxCandidate
import run.nuri.getagrip.runner.SessionOutcome
import run.nuri.getagrip.ui.theme.LocalGripPalette
import java.time.Instant

/// Only reachable through a debug-build intent. Selection never writes to the real store.
@Composable
fun DebugSummaryPreview(onDone: () -> Unit) {
    val outcome = remember {
        val grips = listOf(
            GripSpec(20, FingerSet.four, GripPosition.halfCrimp),
            GripSpec(20, FingerSet.frontThree, GripPosition.halfCrimp),
            GripSpec(20, FingerSet.middleTwo, GripPosition.openHand),
            GripSpec(15, FingerSet.four, GripPosition.fullCrimp),
            GripSpec(20, FingerSet.four, GripPosition.drag),
            GripSpec(45, FingerSet.four, GripPosition.pinch),
        )
        val reps = grips.mapIndexed { i, grip ->
            RepSummary(setIndex = i, grip = grip, side = Side.both, heldSeconds = 10.0,
                peakKg = 12.0 + i, avgKg = 10.0 + i, outcome = RepOutcome.completed)
        }
        val plan = SessionPlan(name = "Daily routine · preview",
            sets = grips.map { SetPlan(grip = it, repsPerSide = 1) }, handMode = HandMode.bothHands)
        SessionOutcome(plan, plan.name, reps, Instant.now().minusSeconds(180), Instant.now(),
            17.0, 12.5, 60.0, 6, 6, false, GaugeKind.progressor, true,
            reps.map { MaxCandidate(it.grip, it.side, it.peakKg, null) }.sortedByDescending { it.kg })
    }
    SessionSummaryScreen(outcome, 1,
        modifier = Modifier.fillMaxSize().background(LocalGripPalette.current.field).safeDrawingPadding()) { _, _ ->
        onDone()
        true
    }
}

File: Sources/Runner/DebugSummaryPreview.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.
#if DEBUG
import SwiftUI
import SwiftData

/// In-memory, interactive QA entry point. No sample logs or maxes reach the real store.
struct DebugSummaryPreview: View {
    @State private var preview = SummaryPreviewState()
    var onDone: () -> Void

    @State private var showLog = ProcessInfo.processInfo.arguments.contains("-previewLog")

    var body: some View {
        SessionSummaryView(template: preview.template, plan: preview.plan, reps: preview.reps,
            startedAt: preview.finishedAt.addingTimeInterval(-180), finishedAt: preview.finishedAt,
            didAnyWork: true, onDone: onDone)
            .sheet(isPresented: $showLog) {
                SessionLogSheet { showLog = false }
            }
            .environment(preview.store)
            .modelContainer(preview.container)
            // Match the production root when auditing the largest supported text.
            .dynamicTypeSize(...DynamicTypeSize.accessibility3)
            .background { AppBackground() }
    }
}

/// Retain the context and its container together across SwiftUI view updates.
@MainActor
private final class SummaryPreviewState {
    let container: ModelContainer
    let template: SessionTemplate
    let plan: SessionPlan
    let reps: [RepSummary]
    let store: TemplateStore
    let finishedAt = Date()

    init() {
        let container = try! ModelContainer(for: SessionTemplate.self, WorkoutLog.self, MaxRecord.self,
            configurations: ModelConfiguration(isStoredInMemoryOnly: true, cloudKitDatabase: .none))
        self.container = container
        let previewHands = ProcessInfo.processInfo.arguments.contains("-previewHandMaxes")
        let defaultGrips = [
            GripSpec(edgeMM: 20, fingers: .four, position: .halfCrimp),
            GripSpec(edgeMM: 20, fingers: .frontThree, position: .halfCrimp),
            GripSpec(edgeMM: 20, fingers: .middleTwo, position: .openHand),
            GripSpec(edgeMM: 15, fingers: .four, position: .fullCrimp),
            GripSpec(edgeMM: 20, fingers: .four, position: .drag),
            GripSpec(edgeMM: 45, fingers: .four, position: .pinch)
        ]
        let grips = previewHands ? Array(repeating: defaultGrips[0], count: 3) : defaultGrips
        var draft = RoutineDraft.blank(named: "Daily routine · preview")
        draft.plan.sets = grips.map { SetPlan(grip: $0, repsPerSide: 1) }
        draft.plan.handMode = .bothHands
        plan = draft.plan
        template = SessionTemplate(draft: draft, sortIndex: 0)
        reps = grips.enumerated().map { index, grip in
            var rep = RepSummary()
            rep.setIndex = index
            rep.grip = grip
            rep.side = previewHands ? [Side.left, .right, .both][index] : .both
            rep.heldSeconds = 10
            rep.peakKg = previewHands ? [35.0, 30.0, 45.0][index] : Double(12 + index)
            rep.avgKg = Double(10 + index)
            rep.outcome = .completed
            return rep
        }
        if previewHands {
            container.mainContext.insert(MaxRecord(grip: grips[0], kg: 40, source: .manual, side: .both))
            try! container.mainContext.save()
        }
        store = TemplateStore(context: container.mainContext,
            clock: DayClock(), settings: SettingsStore(), storageMode: .localOnly)
    }
}
#endif
